src/Orleans/Configuration/GlobalConfiguration.cs

```diff
@@ -670,13 +671,83 @@ internal override void Load(XmlElement root)
+                                existingProviderConfig.Merge(providerConfig);
+                            }
+                            else
+                            {
+                                ProviderConfigurations.Add(providerConfig.Name, providerConfig);
+                            }
                         }
                         break;
                 }
             }
         }
 
+        /// <summary>
+        /// Registers given type of <typeparamref name="T"/> as bootstrap provider
+        /// </summary>
+        /// <typeparam name="T">
+        ///     Non-abstract type which implements <see cref="IBootstrapProvider"/> interface
+        /// </typeparam>
+        /// <param name="properties">
+        ///     Properties that will be passed to bootstrap provider upon initialization
+        /// </param>
+        public void RegisterBootstrapProvider<T>(IDictionary<string, string> properties = null)
+        {
+            RegisterBootstrapProvider(typeof(T), properties);
+        }
+
+        /// <summary>
+        /// Registers given type as bootstrap provider
+        /// </summary>
+        /// <param name="boostrapProviderType">
+        ///     Non-abstract type which implements <see cref="IBootstrapProvider"/> interface
+        /// </param>
+        /// <param name="properties">
+        ///     Properties that will be passed to bootstrap provider upon initialization
+        /// </param>
+        public void RegisterBootstrapProvider(Type boostrapProviderType, IDictionary<string, string> properties = null) 
+        {
+            if (boostrapProviderType == null)
+                throw new ArgumentNullException("boostrapProviderType");
+
+            if (boostrapProviderType.IsAbstract ||
+                boostrapProviderType.IsGenericType ||
+                !typeof(IBootstrapProvider).IsAssignableFrom(boostrapProviderType))
+                throw new ArgumentException("Expected non-generic, non-abstract type which implements IBootstrapProvider interface", 
+                                            "boostrapProviderType");
+
+            const string bootstrapCategoryKey = "Bootstrap";
+
+            ProviderCategoryConfiguration category;
+            ProviderConfigurations.TryGetValue(bootstrapCategoryKey, out category);
+            if (category == null)
+            {
+                category = new ProviderCategoryConfiguration()
+                {
+                    Name = bootstrapCategoryKey,
+                    Providers = new Dictionary<string, IProviderConfiguration>()
+                };
+                ProviderConfigurations.Add(category.Name, category);
+            }
+
+            var fullName = boostrapProviderType.FullName;
+            if (category.Providers.ContainsKey(fullName))
+                throw new InvalidOperationException(
+                    string.Format("Bootstrap provider of type {0} has been already registered", 
```

Comment:
We definitely should not limit to only one impl. of a given Bootstrap Provider. One should be able to pass multiple providers of the exact same type. We currently allow that. The difference in this PR is in the line below:
var config = new ProviderConfiguration(  properties ?? new Dictionary<string, string>(),   fullName, fullName); 
the last arg should be provider name, not fullName of the type. 
It mean the RegisterBootstrapProvider should accept a name.